core/src/main/scala/com/banno/kafka/package.scala

```diff
@@ -179,30 +120,133 @@ package object kafka {
+      * should take care to handle null arguments. A null key means the sender
+      * was unable to choose a key on which to partition. A null value means the
+      * entity identified by the key should be deleted.
+      */
+    override def bitraverse[G[_]: Applicative, A, B, C, D](
         fab: ProducerRecord[A, B]
-    )(f: A => C, g: B => D): ProducerRecord[C, D] =
-      fab.bimap(f, g)
+    )(f: A => G[C], g: B => G[D]): G[ProducerRecord[C, D]] =
+      (f(fab.key), g(fab.value)).mapN(
+        new ProducerRecord(
+          fab.topic,
+          fab.partition,
+          fab.timestamp,
+          _,
+          _,
+          fab.headers,
+        )
+      )
   }
 
-  implicit object ConsumerRecordBifunctor extends Bifunctor[ConsumerRecord] {
-    def bimap[A, B, C, D](
+  implicit object ConsumerRecordBitraverse extends Bitraverse[ConsumerRecord] {
+    override def bifoldLeft[A, B, C](fab: ConsumerRecord[A, B], c: C)(
+        f: (C, A) => C,
+        g: (C, B) => C,
+    ): C =
+      g(f(c, fab.key), fab.value)
+
+    override def bifoldRight[A, B, C](fab: ConsumerRecord[A, B], c: Eval[C])(
+        f: (A, Eval[C]) => Eval[C],
+        g: (B, Eval[C]) => Eval[C],
+    ): Eval[C] =
+      g(fab.value, f(fab.key, c))
+
+    /** Note that since a record's key or value could be null, functions f and g
+      * should take care to handle null arguments. A null key means the sender
+      * was unable to choose a key on which to partition. A null value means the
+      * entity identified by the key should be deleted. Note also that the
+      * returned record has checksum of `-1` since that field is deprecated.
+      */
+    override def bitraverse[G[_]: Applicative, A, B, C, D](
         fab: ConsumerRecord[A, B]
-    )(f: A => C, g: B => D): ConsumerRecord[C, D] =
-      fab.bimap(f, g)
+    )(f: A => G[C], g: B => G[D]): G[ConsumerRecord[C, D]] =
+      (f(fab.key), g(fab.value)).mapN(
+        new ConsumerRecord(
+          fab.topic,
+          fab.partition,
+          fab.offset,
+          fab.timestamp,
+          fab.timestampType,
+          fab.serializedKeySize,
+          fab.serializedValueSize,
+          _,
+          _,
+          fab.headers,
+          fab.leaderEpoch,
+        )
+      )
+  }
+
+  implicit class ByteArrayConsumerRecord(
+      cr: ConsumerRecord[Array[Byte], Array[Byte]]
+  ) {
+    def maybeKeyAs[K](topic: String)(implicit kd: Deserializer[K]): Option[K] =
+      cr.maybeKey.map(kd.deserialize(topic, _))
+    def maybeValueAs[V](topic: String)(implicit
+        vd: Deserializer[V]
+    ): Option[V] =
+      cr.maybeValue.map(vd.deserialize(topic, _))
+
+    // note that these will probably throw NPE if key/value is null
+    def keyAs[K](topic: String)(implicit kd: Deserializer[K]): K =
+      kd.deserialize(topic, cr.key)
+    def valueAs[V](topic: String)(implicit vd: Deserializer[V]): V =
+      vd.deserialize(topic, cr.value)
+
+    /** This only works when both key and value are non-null. */
```

Comment:
Same thing here, if this throws exceptions, we should probably communicate that to the caller by returning `Try[ConsumerRecord[K, V]]`.

Reply:
As https://github.com/Banno/kafka4s/pull/720#discussion_r1089466023